### PR TITLE
Migrate TextInputPlugin API to TextRange

### PR DESCRIPTION
**Pull request by cbracken**
## Description

Replaces selection_base() and selection_extent() with selection() and SetSelection(int, int) with SetSelection(range).

This also adds the following convenience methods to TextRange:
* reversed()
* Contains(size_t position)
* Contains(const TextRange& range)

as well as operator== for use in unit tests. When Flutter migrates to C++20, we can replace that method with a default declaration.

## Related Issues

Full IME support for Windows (flutter/flutter#65574)
Full IME support for Linux (flutter/flutter#66880)

## Tests

I added the following tests:

* Tests for all new methods on TextRange
* Existing tests for TextInputModel were updated for the new API

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [ ] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

File: shell/platform/common/cpp/text_range.h [hunk 1]
```diff
@@ -47,6 +47,23 @@ class TextRange {
...
 
+  // Returns true if the base is greater than the extent.
+  bool reversed() const { return base_ > extent_; }
+
+  // Returns true if |position| is contained within the range.
+  bool Contains(size_t position) const {
+    return position >= start() && position < end();
+  }
+
+  // Returns true if |position| is contained within the range.
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/text_range.h:58]: s/position/range/

File: shell/platform/common/cpp/text_range.h [hunk 2]
```diff
@@ -47,6 +47,23 @@ class TextRange {
   // Returns true if the range is of length 0.
   bool collapsed() const { return base_ == extent_; }
 
+  // Returns true if the base is greater than the extent.
+  bool reversed() const { return base_ > extent_; }
+
+  // Returns true if |position| is contained within the range.
+  bool Contains(size_t position) const {
+    return position >= start() && position < end();
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/text_range.h:55]: Why is this non-inclusive, but the `range` version below is inclusive? Doesn't this mean that `Contains(some_collapsed_range)` and `Contains(some_collapsed_range.position())` would return different results?

File: shell/platform/common/cpp/text_range.h (hunk 1 above)
Comment (cbracken) [reply shell/platform/common/cpp/text_range.h:58]: Done.

File: shell/platform/common/cpp/text_range.h (hunk 2 above)
Comment (cbracken) [reply shell/platform/common/cpp/text_range.h:55]: Thanks for spotting this. I'd originally written them both as <, realised we need to cover cursor positions after the laste text position, then went back and fixed the range-based version but didn't fix the position version or the tests. Fixed now.

**Comment by cbracken** [comment]
Added four additional tests covering ranges that span the start/end positions + the reversed versions.

**Comment by cbracken** [comment]
Linux Android Scenarios failure is an infra flake. Kicked the bot. New run: https://ci.chromium.org/p/flutter/builders/try/Linux%20Android%20Scenarios/2566

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Comment by cbracken** [comment]
Top row of engine build is now green; landing.

**Pull request merged**